cosmwasm-plugin/main.go

```diff
@@ -0,0 +1,296 @@
+package main
+
+import (
+	"embed"
+	"encoding/gob"
+	"fmt"
+	"io"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	hplugin "github.com/hashicorp/go-plugin"
+
+	"github.com/ignite/cli/ignite/services/chain"
+	"github.com/ignite/cli/ignite/services/plugin"
+)
+
+var (
+	chainName  string
+	chainNamed string
+	//go:embed wasm-wiring/*
+	templates embed.FS // Embedded template files
+
+)
+
+func init() {
+	gob.Register(plugin.Manifest{})
+	gob.Register(plugin.ExecutedCommand{})
+	gob.Register(plugin.ExecutedHook{})
+}
+
+type p struct{}
+
+func (p) Manifest() (plugin.Manifest, error) {
+	return plugin.Manifest{
+		Name: "cosmwasm-plugin",
+		// Add commands here
+		Commands: []plugin.Command{
+			// Example of a command
+			{
+				Use:   "cosmwasm-plugin",
+				Short: "Explain what the command is doing...",
+				Long:  "Long description goes here...",
+				Flags: []plugin.Flag{
+					{Name: "my-flag", Type: plugin.FlagTypeString, Usage: "my flag description"},
+				},
+				PlaceCommandUnder: "ignite",
+				// Examples of adding subcommands:
+				Commands: []plugin.Command{
+					{
+						Use:               "add",
+						Short:             "Add a CosmWasm module to the chain",
+						Long:              "This command will install dependencies, create new files, and modify existing files to add a CosmWasm module to the chain.",
+						PlaceCommandUnder: "ignite cosmwasm-plugin",
+					},
+				},
+			},
+		},
+		// Add hooks here
+		Hooks:      []plugin.Hook{},
+		SharedHost: false,
+	}, nil
+}
+
+func (p) Execute(cmd plugin.ExecutedCommand) error {
+
+	fmt.Printf("Hello I'm the cosmwasm-plugin\n")
+
+	// This is how the plugin can access the chain:
+	// c, err := getChain(cmd)
+
+	// According to the number of declared commands, you may need a switch:
+
+	switch cmd.Use {
+	case "add":
+		return handleAddCommand()
+	default:
+		return fmt.Errorf("unknown command: %s", cmd.Use)
+	}
+}
+
+func handleAddCommand() error {
+	fmt.Println("Adding stuff...")
+
+	err := installDependencies()
+	if err != nil {
+		return err
+	}
+
+	err = createNewFiles()
+	if err != nil {
+		return err
+	}
+
+	//err = modifyExistingFiles()
+	//if err != nil {return err}
+
+	return nil
+}
+
+func (p) ExecuteHookPre(hook plugin.ExecutedHook) error {
+	fmt.Printf("Executing hook pre %q\n", hook.Name)
+	return nil
+}
+
+func (p) ExecuteHookPost(hook plugin.ExecutedHook) error {
+	fmt.Printf("Executing hook post %q\n", hook.Name)
+	return nil
+}
+
+func (p) ExecuteHookCleanUp(hook plugin.ExecutedHook) error {
+	fmt.Printf("Executing hook cleanup %q\n", hook.Name)
+	return nil
+}
```

Comment:
I think these three hook methods could be removed if you define the plugin as:

```go
type p struct{
	plugin.Interface
}
```

Otherwise code could be simplified:

```suggestion
func (p) ExecuteHookPre(plugin.ExecutedHook) error {
	return nil
}

func (p) ExecuteHookPost(plugin.ExecutedHook) error {
	return nil
}

func (p) ExecuteHookCleanUp(plugin.ExecutedHook) error {
	return nil
}
```

Reply:
Implemented change with [commit](https://github.com/ignite/plugins/commit/f5cc05c1b4ee6d6b5e642868252b1ab2e0e74d8e)